libsolidity/analysis/ControlFlowAnalyzer.h

```diff
@@ -33,16 +33,26 @@ class ControlFlowAnalyzer: private ASTConstVisitor
 	bool analyze(ASTNode const& _astRoot);
 
 	bool visit(FunctionDefinition const& _function) override;
+	bool visit(ContractDefinition const& _contract) override;
 
 private:
+	void analyze(FunctionDefinition const& _function, ContractDefinition const* _contract = nullptr);
 	/// Checks for uninitialized variable accesses in the control flow between @param _entry and @param _exit.
-	void checkUninitializedAccess(CFGNode const* _entry, CFGNode const* _exit, bool _emptyBody) const;
+	void checkUninitializedAccess(CFGNode const* _entry, CFGNode const* _exit, bool _emptyBody, FunctionDefinition const& _function, ContractDefinition const* _contract);
 	/// Checks for unreachable code, i.e. code ending in @param _exit, @param _revert or @param _transactionReturn
 	/// that can not be reached from @param _entry.
-	void checkUnreachable(CFGNode const* _entry, CFGNode const* _exit, CFGNode const* _revert, CFGNode const* _transactionReturn) const;
+	void checkUnreachable(CFGNode const* _entry, CFGNode const* _exit, CFGNode const* _revert, CFGNode const* _transactionReturn);
+
+	bool checkForReverts(ContractDefinition const& _contract, FunctionCall const& _functionCall);
 
 	CFG const& m_cfg;
 	langutil::ErrorReporter& m_errorReporter;
+
+	std::set<langutil::SourceLocation> m_previousUnreachable;
```

Comment:
Most of these members are only used during the revert analysis, so I think it would be nice to extract them into their own class. After the revert analysis, the findings should be stored in the CFG.

Reply:
The fact whether or not a function always reverts should be stored in `FunctionFlow`, shouldn't it?